---
title: "Single Cell RNAseq "
author: "Chenling Xu Antelope"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Vignettes are long form documentation commonly included in packages. 

### load requred packages and pre-computed simulations and data-matched parameters
```{r, fig.show='hold',message=F}
setwd('/Users/chenlingantelope/ResearchProjects/scRNA_simulation')
required_packages <- c('devtools','FNN','reshape','ggplot2','gridExtra','ape','MASS','Rtsne','RColorBrewer','fAsianOptions','fOptions','fBasics','timeSeries','timeDate','Biobase' )
install_packages <- rownames(installed.packages())
need_install <- setdiff(required_packages,install_packages)
if(length(need_install)>0){install.packages(need_install)}

load_result <- lapply(required_packages, require, character.only = TRUE)

load('allsim.20170706.robj')
load('match_params.robj')
```
### load package SCsim and set up plotting size in jupyter
```{r,message=F}
setwd('/Users/chenlingantelope/ResearchProjects/scRNA_simulation')
load_all('SCsim/')
ngenes=100
ncells=100
nevf=10
randseed=0
```

# Overview
The input parameters of the package allow users to control intrinsic variation, technical variation and biological variation in the data. Intrinsic variation is modeled through the kinetic model; technical variation takes into account dropouts, batch effects, GC and length biases of genes; and biological variation is modeled by EVFs (extrinsic variation factors). 

**sim1Pop1Batch** is the main function in the SCsim package. It takes all of the parameters needed for generating the count matrix for a homogeneous population that is sequenced in 1 batch.

 The input parameters are listed below while being categorized by the level of variation they contribute to:

Parameters that affect intrinsic variation:
* **bimod** the proportion of distance to move a point in the kon-koff space towards the center of the kon-koff space where the gene expression is bimodal

Parameters that affect technical variation:
* **alpha** library prep capture efficiency, or dropout of transcript: the number of of captured transcript is distributed as the number of successes with probability of success alpha in X trials, where X is the true number of transcripts
* **alpha_sd** the standard deviation of alpha (to add noise for different efficiency for each cell)
* **nbins** the number of bins for gc and length bias
* **gcbias** the magnitude of gc bias (a number between 0 and 2) 
* **lenbias** the magnitude of length bias (a number between 0 and 2)
* **batch_mean** The mean batch effect from which the batch effect of each gene is sampled from
* **batch_sd** The standard deviation of batch effect from which the batch effect of each gene is sampled from
* **noise** the standard deviation of a normal distribution where the log(noise) is sampled from
* **randseed** the random seed to generate the gc and length bins (should keep constant for each experiment regardless of the batch and population, but should change for replicates of simulations)

Parameters that affect biological variation:
* **evf_mean** a vector of length nevf, the means of the normal distribution to sample EVF from 
* **evf_sd** a vector of length nevf, the standard deviation of the normal distribution to sample EVF from
* **gene_effects** a matrix of ngenes * nevf, generated by GeneEffects function

Other parameters: 
* **ncells** number of cells to simulate


# Prepare input parameters: Compute gene effects 
The gene effect is a matrix of ngenes*nevf, generated by GeneEffects function and remains the same across all experiments. Its role is explained in the **evf** section of this document
```{r,cache=F}
gene_effects <- GeneEffects(ngenes=ngenes,nevf=10,randseed=0,sd=1,prob=0.3)
```

# Simulation: an example run 
We will simulate a dataset using the sim1Pop1Batch and visualize the results with function PlotCountHeatmap. The output of the sim1Pop1Batch function is a list of sampled EVF for each cell, and transcript count matrices that have different levels of noise added to it. In the follwing example, temp is the output of the funcion. temp[[1]] is a list of **evf** for each cell, temp[[2]] is the true transcript counts directly sampled from the stationary distribution given expression dynamic parameters, temp[[3]] is after dropout, and temp[[4]] is the dropped out counts plus batch effects and gene-specific biases.

```{r,fig.height = 3, fig.width = 10, fig.align = "center",cache=F}
temp<-sim1Pop1Batch(evf_mean=rep(1,nevf), evf_sd=rep(0.2,nevf),ncells=ncells,
        randseed=0,gene_effects=gene_effects,bimod=0.1,alpha=0.1,
        alpha_sd=0.01,nbins=10,gcbias=0.1,lenbias=0.1,batch=exp(rnorm(ngenes,0,0.1)),noise=0.1,
        matched_params=match_params1)
log_count_bins<-seq(0,4,0.4)
knitr::kable(head(do.call(rbind,temp[[1]]), 2))
par(mfrow = c(1,3))
ord1<-PlotCountHeatmap(LogDist(temp[[2]],log_count_bins),rowMeans(temp[[2]]),NA,1,paste('truecounts.bimod=',bimod,'.jpeg',sep=''))
ord2<-PlotCountHeatmap(LogDist(temp[[3]],log_count_bins),rowMeans(temp[[3]]),NA,1,paste('droppedcounts.bimod=',bimod,'.jpeg',sep=''))
ord3<-PlotCountHeatmap(LogDist(temp[[4]],log_count_bins),rowMeans(temp[[4]]),NA,1,paste('biasedcounts.bimod=',bimod,'.jpeg',sep=''))
gridExtra::grid.arrange(ord1,ord2,ord3,ncol=3)
```
# Test the effect of parameters on changing various levels of variation
## The intrinsic variation from the stochastic nature of gene expression
We use a kinetic gene expression model that assumes that a gene switches between an 'on' state and an 'off' state. The rate of turning on is $k_{on}$ and the rate of turning off is $k_{off}$. Transcripts are synthesized at rate $s$ when the gene is on. They also degrade at rate $d$ regardless of the state. From these four parameters we can calculate the stationary distribution of the abundance of each transcript in each cell. The first variation that we simulate is this intrinsic variation, which means that even when expression is governed by exactly the same parameters (or as we will refer to it, the cells sharing the same parameters are in the same population), cells might have different transcript counts. The amount of variation can be tuned by changing the parameter range. 

We give each gene in the same population the same $k_{on}$, $k_{off}$, $s$ and $d$ ($d$ is always 1) with some noise added for each cell. We noticed that the expression distribution tend to be bimodal, and has high variation when $k_{on}$ and $k_{off}$ are very small, as shown on the heatmap below. The first plot shows how the fano factor increases at the bottom left corner when $s=10$. To increase the intrinsic variation we can vary the bimod, or $\beta$ parameter so that the $k_{on}$ and $k_{off}$ parameter is closer to the bottom left corner within the region of bimodal distribution (purple clolor in the second plot, where the value 2 indicates that the distribution has 2 inflection points). Note that the axis are on log scale. 






































You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
